Support counters with tablets

mlitvk opened the pull request:
Support the counters feature in tablets keyspaces.

The main change is to fix the counter update during tablets intranode migration.

Counter cell is c = map<host_id, value>. A counter update is applied by doing read-modify-write on a leader replica to retrieve the current host's counter value and transform the mutation to contain the updated value for the host, then apply the mutation and replicate it to other hosts. the read-modify-write is protected against concurrent updates by locking the counter cell.

When the counter is migrated between two shards, it's not enough to lock the counter on the read shard, because in the stage write_both_read_new the read shard is switched, and then we can have concurrent updates reach either the old or the new shard. In order to keep the counter update exclusive we lock both shards when in the stage write_both_read_new.

Also, when applying the transformed mutation we need to respect write_both stages and apply the mutation on both shards. We change it to use `apply_on_shards` similarly to other methods in storage proxy.

The change applies to both tablets and vnodes, they use the same implementation, but for vnodes the behavior should remain equivalent up to some small reordering of the code since it doesn't have intranode migration and reduces to single read shard = write shard.

Fixes https://github.com/scylladb/scylladb/issues/18180

no backport - new feature

gleb-cloudius commented:
@gusev-p do you remember if we have fencing for counters?

mlitvk commented:
> How is fencing managed? For regular reads/writes, we propagate the fence from the coordinator to the replicas. Here, we have to propagate the fence from the coordinator to the leader to the replicas. Are the relevant RPCs enriched with fencing?
> 

I'm not sure because I'm not really familiar with the fencing, but I see we have a fence covering `mutate_counters_on_leader`, and it's also sent in the RPC to the leader, and it also covers the replication from leader part
```cpp
            co_await apply_fence_on_ready(this->mutate_counters_on_leader(std::move(endpoint_and_mutations.second), cl, timeout, tr_state, permit), fence, my_address);
```

gusev-p commented on replica/database.cc:
```diff
@@ -2054,29 +2026,65 @@ future<> database::apply_in_memory(const mutation& m, column_family& cf, db::rp_
+
+    tracing::trace(trace_state, "Acquiring counter locks");
+
+    co_return co_await update_write_metrics([&m, &cf, op = std::move(op), timeout] mutable -> future<counter_update_guard> {
```
`update_write_metrics` doesn't measure the write operation, we now increment `total_writes` three times for a single counter write -- in `acquire_counter_locks`, `prepare_counter_update` and `apply_counter_update`.

mlitvk commented on replica/database.cc:
```diff
@@ -1910,7 +1910,7 @@ static query::partition_slice partition_slice_for_counter_update(const mutation&
 }
 
-future<> database::read_and_transform_counter_mutation_to_shards(mutation& m, column_family& cf,
+future<mutation> database::read_and_transform_counter_mutation_to_shards(mutation m, column_family& cf,
```
I thought it's moved when we call it and when we return so it shouldn't matter much
`m` is local in the caller function, and previously the caller was a coroutine so we could pass it by reference. here the caller is changed to a non-coroutine so I move it instead

mlitvk replied:
I checked it with a simple test and I don't see that it's copied
```
$ cat test_return.cc 
#include <iostream>
#include <seastar/core/future.hh>
#include <seastar/core/coroutine.hh>

struct TestValue {
    TestValue() = default;
    TestValue(const TestValue&) { std::cout << "Copied\n"; }
    TestValue(TestValue&&) noexcept { std::cout << "Moved\n"; }
    TestValue& operator=(const TestValue&) { std::cout << "Copy assigned\n"; return *this; }
    TestValue& operator=(TestValue&&) noexcept { std::cout << "Move assigned\n"; return *this; }
};

seastar::future<TestValue> test_coroutine(TestValue v) {
    co_return v;
}

int main() {
    TestValue v;
    auto fut = test_coroutine(std::move(v));
    auto result = fut.get();
    return 0;
}
$ g++ test_return.cc $(pkg-config --libs --cflags --static ./seastar/build/release/seastar.pc) -o test_return
$ ./test_return 
Moved
Moved
Moved
Moved
```

mlitvk replied:
added `std::move`

gusev-p commented on test/cqlpy/cassandra_tests/batch_test.py:
```diff
@@ -44,39 +44,21 @@ def testNonCounterInCounterBatch(cql, test_keyspace):
 
-@pytest.mark.parametrize("test_keyspace",
-                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #18180")]), "vnodes"],
-                         indirect=True)
```
I was under impression that we already have a parameter for vnodes/tablets switch in test.py, but it seems we don't. @temichus maybe your team can help? The idea is to run test.py tests for vnodes in a separate Jenkins job, like we do for dtests. We need to add a test.py parameter which would be used by the `test_keyspace` fixture and migrate explicit parameterizations (`parametrize("test_keyspace", ["tablets", "vnodes"], indirect=True)`) to it.

mlitvk commented:
* rebased and fixed conflicts with `run_fenceable_write`
* addressed comments: added `std::move` and changed to destroy the guard implicitly by scope

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -3582,33 +3582,62 @@ storage_proxy::mutate_counters_on_leader(utils::chunked_vector<frozen_mutation_a
-                return proxy.local().replicate_counter_from_leader(std::move(m), cl, std::move(trace_state), timeout, std::move(p));
+    const auto& rs = s->table().get_effective_replication_map()->get_replication_strategy();
+
+    auto m = co_await run_fenceable_write(rs, fence, caller, [this, erm, s, fm = std::move(fm), timeout, trace_state] (this auto) -> future<mutation> {
```
why we need `(this auto)` parameter?

piodul commented on test/cluster/test_replica_exceptions.py:
```diff
@@ -152,7 +152,7 @@ async def test_replica_writes_do_apply_counter_update_timeout(manager: ManagerCl
 
     measurement = (m := Measurement())._replace(
-        cpp_exception_threshold = 20 + m.run_count * 10,
+        cpp_exception_threshold = 20 + m.run_count * 15,
```
I think some of the coroutinizations that you have made might have done this. So, the test does its job and you should fix the exception propagation logic on the hot paths.

For example:

```c++
    auto m = co_await run_fenceable_write(rs, fence, caller, [this, s, fm = std::move(fm), timeout, trace_state] {
            return _db.local().apply_counter_update(s, fm, timeout, trace_state);
        });
```

If `run_fenceable_write` fails with an exception, the `co_await` operator will rethrow it. On the hot path, you should be careful to use variants which do not throw. So, for example:

```c++
auto f_m = co_await coroutine::as_future(run_fenceable_write(rs, fence, caller, [this, s, fm = std::move(fm), timeout, trace_state] {
            return _db.local().apply_counter_update(s, fm, timeout, trace_state);
        });

if (f_m.failed()) {
    co_await coroutine::return_exception(f_m.get_exception());
}
auto m = f.get();

// ...
```

Every `co_await` should be scrutinized. It sucks, but that's unfortunately a drawback of the coroutines - continuations are less throw-happy.

mlitvk commented on service/storage_proxy.cc:
```diff
@@ -3599,13 +3599,29 @@ future<>
+        std::ranges::sort(lock_shards);
+
+        using foreign_counter_guard = foreign_ptr<lw_shared_ptr<replica::counter_update_guard>>;
+        std::vector<foreign_counter_guard> counter_locks;
```
done
